### PR TITLE
Leaves - Kristina and Julia

### PR DESCRIPTION
**Pull request by Kalakalot**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? | the signature, the parameter(s), and if you're calling a method, the arguments
What are the advantages of using git when collaboratively working on one code base? | Git provides version control and allows you to add notes when you are making changes, which is helpful for your team members (and yourself). Git also shows who made which changes, and provides ability to compare differences. 
What kind of relationship did you and your pair have with the unit tests? | A challenging one: it took us a while to figure out how the tests fit with our code and the driver code, but once we understood, tests made finding errors and pinpointing what was wrong much easier. 
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We successfully used .max to find the highest-scoring word. We tried to implement .select to help us with tie-breaking for equally scoring words (wave 4), but we ran out of time and brainpower. 
What was one method you and your pair used to debug code? | We used the errors and stack trace from tests to determine where errors in our code originated. 
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We talked about learning styles (very similar) and we also checked in about emotions frequently, which helped us understand where the other person was coming from. We bonded over lack of sleep and bad but funny jokes.

### CODE REVIEW

File: adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+
+# Wave one method
+def draw_letters
+  remaining_pool = %w(a a a a a a a a a b b c c d d d d e e e e e e e e e e e e f f g g g h h i i i i i i i i i j k l l l l m m n n n n n n o o o o o o o o p p q r r r r r r s s s s t t t t t t u u u u v v w w x y y z)
+  remaining_pool.map!(&:upcase)
+  letters_in_hand = []
+  
+  10.times do |letter|
+    letter = remaining_pool.sample
```
Comment (dHelmgren) [inline adagrams.rb:14]: There is another version of sample! It might help you DRY up this code.

File: adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,148 @@
+require 'pry'
+
+letter_pool = %w(a a a a a a a a a b b c c d d d d e e e e e e e e e e e e f f g g g h h i i i i i i i i i j k l l l l m m n n n n n n o o o o o o o o p p q r r r r r r s s s s t t t t t t u u u u v v w w x y y z)
+
+
+
+# Wave one method
+def draw_letters
+  remaining_pool = %w(a a a a a a a a a b b c c d d d d e e e e e e e e e e e e f f g g g h h i i i i i i i i i j k l l l l m m n n n n n n o o o o o o o o p p q r r r r r r s s s s t t t t t t u u u u v v w w x y y z)
```
Comment (dHelmgren) [inline adagrams.rb:9]: Why do we redo the work of line 3 here?

File: adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    remaining_pool.slice!(remaining_pool.index(letter))
+  end 
+  # p remaining_pool
+  return letters_in_hand
+end 
+
+
+
+# Wave 2: uses_available_letters?
+# Need to review lines 57-67, b/c failing test for returning true if the submitted letters are valid against the drawn letters
```
Comment (dHelmgren) [inline adagrams.rb:25]: This comment should come out before submission!

File: adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+def uses_available_letters?(input, letters_in_hand)
+  
+  # Make input word uppercase and then split it into an array
+  input_array = input.upcase.chars
+  
+  # Check if the input string's characters exist in the user hand 
+  
+  input_array.each do |letter|
+    unless letters_in_hand.include?(letter) 
+      return false
```
Comment (dHelmgren) [inline adagrams.rb:35]: You end up repeating this work down below.

File: adagrams.rb [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    if letters_hand_hash.key?(found_letter)
+      letters_hand_hash[found_letter] += 1
+    else
+      letters_hand_hash[found_letter] = 1
+    end 
+  end 
+  
+  # making a hash from user_input with letters as keys and counts as values 
+  input_hash = {}
+  input_array.each do |letter|
```
Comment (dHelmgren) [inline adagrams.rb:55]: I think making a second hash complicates this a lot. What if you just used the input string to count down the number of letters from `letters_hand_hash`?

File: adagrams.rb [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  end 
+  
+  # Calculating the sum for all the letters in word 
+  sum = 0
+  scores.each do |score|
+    sum = sum + score
+  end 
+  
+  # If the length of word is between 7-10, add 8 points to the total score 
+  if input_array.length == 7 || input_array.length == 8 || input_array.length == 9 || input_array.length == 10
```
Comment (dHelmgren) [inline adagrams.rb:110]: You can reduce the amount of work you are doing here! It's possible to do this line with only 2 comparisons. What are they?

<details>
<code> if input_array.length >= 7  && input_array.length <= 10
</details>

File: adagrams.rb [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+      scores.push(8)
+    elsif word == "Q" || word =="Z"
+      scores.push(10)
+    end 
+  end 
+  
+  # Calculating the sum for all the letters in word 
+  sum = 0
+  scores.each do |score|
+    sum = sum + score
```
Comment (dHelmgren) [inline adagrams.rb:106]: If you're trying to end up with a sum, why make an array and then convert that array into a bunch of letters? You can just add up all the letter scores as you go!

File: adagrams.rb [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+def highest_score_from(words)
+  # "words" is an array of strings that is provided via driver code
+  word_scores = [] #contains score for each word
+  word_lengths = [] #contains lengths for each word
+  
+  # need to either define or get rid of "word"
+  
+  words.each do |word|
+    word_letters = word.chars #convert string to array of characters
+    word_length = word_letters.length #count the length of characters
```
Comment (dHelmgren) [inline adagrams.rb:128]: You added a step here. `word.length` is a valid string operation!

File: adagrams.rb [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    word_length = word_letters.length #count the length of characters 
+    word_lengths.push(word_length) #add the length of the word into word_lengths list
+    #calculate the score for each word
+    final_score = score_word(word) 
+    word_scores.push(final_score) #add the score for each word in a list
+  end 
+  
+  # determine which word has the highest score and return associated word
+  highest_score = word_scores.max
+
```
Comment (dHelmgren) [inline adagrams.rb:137]: You currently have 2 failing tests. The code that will fix them goes here. You need to either 1) prefer the shortest word or 2) prefer the word that uses all 10 letters, depending on input.

File: lib/adagrams.rb [hunk 10]
```diff
@@ -0,0 +1,174 @@
+require 'pry'
+
+letter_pool = %w(a a a a a a a a a b b c c d d d d e e e e e e e e e e e e f f g g g h h i i i i i i i i i j k l l l l m m n n n n n n o o o o o o o o p p q r r r r r r s s s s t t t t t t u u u u v v w w x y y z)
```
Comment (dHelmgren) [inline lib/adagrams.rb:3]: looks like you have two versions of this file. was that intentional?

**Comment by dHelmgren** [comment]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | Yes
Small commits with meaningful commit messages | yes
**Code Requirements** |
`draw_letters` method |
Uses appropriate data structure to store the letter distribution | yes, but see comments
All tests for `draw_letters` pass | yes
`uses_available_letters?` method |
All tests for `uses_available_letters?` pass | yes
`score_word` method |
Uses appropriate data structure to store the letter scores | yes
All tests for `score_word` pass | No, see comments. 
`highest_score_from` method |
Appropriately handles edge cases for tie-breaking logic | yes
All tests for `highest_score_from` pass | yes
**Overall** | Good work on this, though some of your tests are failing. I left a few comments around this code. Your process right now is solid, and has gotten you to many correct answers. However, you end up repeating your work a bunch, or doing things with more lines of code than you need. Checking in with the documentation to find just the right method is a really clutch skill that can sometimes save hours of work. As we progress in the curriculum, focus on making sure your fundamentals are sharp, and that you check and recheck your mental toolbox for the right built-in tools.
